internal/charmstore/store.go

```diff
@@ -951,31 +936,96 @@ func (s *Store) SetPromulgated(url *router.ResolvedURL, promulgate bool) error {
 	// select all entities with a matching promulgated URL name. Because of
 	// 2) we are sure that we are only updating all charms or the single
 	// bundle entity.
+
+	iter = s.DB.Entities().Find(bson.D{{
+		"promulgated-revision", bson.D{{"$gt", -1}},
+	}, {
+		"name", base.Name,
+	}}).Select(FieldSelector("promulgated-revision", "supportedseries", "series")).Iter()
+
 	latestPromulgated := make(map[string]int)
+	oldMultiSeries := false
+	var e mongodoc.Entity
+	for iter.Next(&e) {
+		oldMultiSeries = oldMultiSeries || e.Series == ""
+		entitySeries := e.SupportedSeries
+		if e.Series == "bundle" {
+			entitySeries = []string{"bundle"}
+		}
+		for _, series := range entitySeries {
+			if rev, ok := latestPromulgated[series]; !ok || rev < e.PromulgatedRevision {
+				latestPromulgated[series] = e.PromulgatedRevision
+			}
+		}
+	}
+	if err := iter.Err(); err != nil {
+		return errgo.Notef(err, "cannot close mgo iterator")
+	}
+
+	// Find the latest revision in each series of entities with the promulgated base URL.
+	// After this, latestOwned will have an entry for each series, with multi-series
+	// charms having an empty series.
+	type result struct {
+		URL             *charm.URL
+		Series          string `bson:"_id"`
+		SupportedSeries []string
+		Revision        int
+	}
+	latestOwned := make(map[string]result)
 	iter = s.DB.Entities().Pipe([]bson.D{
-		{{"$match", bson.D{{"name", base.Name}}}},
-		{{"$group", bson.D{{"_id", "$series"}, {"revision", bson.D{{"$max", "$promulgated-revision"}}}}}},
+		{{"$match", bson.D{{"baseurl", base}}}},
+		{{"$sort", bson.D{{"revision", 1}}}},
+		{{"$group", bson.D{
+			{"_id", "$series"},
+			{"url", bson.D{{"$last", "$_id"}}},
+			{"supportedseries", bson.D{{"$last", "$supportedseries"}}},
+			{"revision", bson.D{{"$last", "$revision"}}},
+		}}},
 	}).Iter()
-	var res result
-	for iter.Next(&res) {
-		latestPromulgated[res.Series] = res.Revision
+	var r result
+	for iter.Next(&r) {
+		latestOwned[r.Series] = r
 	}
-	if err := iter.Close(); err != nil {
+	if err := iter.Err(); err != nil {
 		return errgo.Notef(err, "cannot close mgo iterator")
 	}
 
-	// Update the newest entity in each series with a base URL that matches the newly promulgated
-	// base entity to have a promulgated URL, if it does not already have one.
+	// Delete all series we don't want to promulgate.
+	if _, ok := latestOwned[""]; ok || oldMultiSeries {
+		// There's a new multi-series charm or there were any old
```

Comment:
// There was previously a multi-series charm, or the newly promulgated 
// charm will be multi-series, in which case we do not promulgate any single 
// series charms. ?